src/network/core/packet.cpp

```diff
@@ -293,16 +352,71 @@ void Packet::Recv_string(char *buffer, size_t size, StringValidationSettings set
 	if (cs->HasClientQuit()) return;
 
 	pos = this->pos;
-	while (--size > 0 && pos < this->size && (*buffer++ = this->buffer[pos++]) != '\0') {}
+	while (--size > 0 && pos < this->Size() && (*buffer++ = this->buffer[pos++]) != '\0') {}
 
-	if (size == 0 || pos == this->size) {
+	if (size == 0 || pos == this->Size()) {
 		*buffer = '\0';
 		/* If size was sooner to zero then the string in the stream
 		 *  skip till the \0, so than packet can be read out correctly for the rest */
-		while (pos < this->size && this->buffer[pos] != '\0') pos++;
+		while (pos < this->Size() && this->buffer[pos] != '\0') pos++;
 		pos++;
 	}
 	this->pos = pos;
 
 	str_validate(bufp, last, settings);
 }
+
+/**
+ * Get the amount of bytes that are still available for the Transfer functions.
+ * @return The number of bytes that still have to be transfered.
+ */
+size_t Packet::RemainingBytesToTransfer() const
+{
+	return this->Size() - this->pos;
+}
+
+/**
+ * Transfer data from the packet to the given function. It starts reading at the
+ * position the last transfer stopped.
+ * See Packet::TransferIn for more information about transferring data to functions.
+ * @param transfer_function The function to pass the const char *buffer and size to write,
+ *                          which returns the amount that was accepted or -1 upon errors.
+ * @return The return value of the transfer_function.
+ */
+ssize_t Packet::TransferOut(std::function<ssize_t(const char*, size_t)> transfer_function, size_t limit)
+{
+	size_t amount = std::min(this->RemainingBytesToTransfer(), limit);
+	if (amount == 0) return 0;
+
+	const char *output_buffer = this->buffer.data() + this->pos;
+	ssize_t bytes = transfer_function(output_buffer, amount);
+	if (bytes > 0) this->pos += bytes;
+	return bytes;
+
+}
+
+/**
+ * Transfer data from the given function into the packet. It starts writing at the
+ * position the last transfer stopped.
+ *
+ * With transfering data from (or to) a function you have to about TCP's recv and
```

Comment:
`transfering` -> `transferring`.
But this sentence seems to be missing a word? I can't make sense out of it, sorry :)